helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java

```diff
@@ -838,7 +839,7 @@ private static void addCompletedTasks(Set<Integer> set, JobContext ctx, Iterable
    */
   private boolean isTaskNotInTerminalState(TaskPartitionState state) {
```

Comment:
This change also impacts filterTasks(). The logic looks to be valid even after this change.
The concerning part is that there are some duplicate checks here and there. For example,

      // Allow tasks eligible for scheduling
      if (state == null || state == TaskPartitionState.STOPPED
          || state == TaskPartitionState.TIMED_OUT || state == TaskPartitionState.TASK_ERROR
          || state == TaskPartitionState.DROPPED) {
        filteredTasks.add(partitionNumber);
      }
      // Allow tasks whose assigned instances are no longer live for rescheduling
      if (isTaskNotInTerminalState(state)) {
        ...
      }

This section checks for different state combinations and then react differently. First of all, it is hard for the reviewers to follow the logic. Secondly, it is very easy for the coder to introduce bugs since the combinations are separately done in different private methods.
Is it possible that we conclude all the states check within 3 or 4 methods? Or we need to clean up the state. The current design is not reviewable.

Reply:
We need to note that this no design. This is a bug which is introduced before and is impacting production result. I tried differently but I believe this is the best way.

Reply:
Also here we are trying to minimize the usage of TaskPartitionState.... in this function and move the logic to other functions. If not, the code will be ugly and hard to follow. By just using given up, we mean this task is not reliable by the controller.

Reply:
One more thing to add. I think the designer of this method (I was not involved in designing this method BTW) used the variable `partitionsToRetryOnLiveInstanceChangeForTargetedJob` to give some hint to the reader about what are these tasks and why we are adding them back. Hence, I think keeping it as it is still the best solution.